# Student Performance Indicator

### Life cycle of a machine learning project
- Understanding the problem statement
- Data collection
- Data checks to perform
- Exploratory Data Analysis (EDA)
- Data pre-processing
- Model training
- Choose best model

## 1. Problem Statement
This project seeks to understand how student's performance, measured by test scores, is affected by other variables such as Gender, Ethnicity, Parental Level of Education, whether the student receives lunch at school, and whether the student participated in a test preparation course.

## 2. Data Collection
Dataset Source: [Kaggle - Student Performance Dataset](https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?resource=download&select=StudentsPerformance.csv)

The dataset consists of 8 columns and 1000 rows

### 2.1 Import Data and Required Packages
Importing Pandas, NumPy, Matplotlib, Seaborn, and Warnings Library.

In [54]:
pip install pyjanitor

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 123.1 kB/s  0:00:09137.1 kB/s eta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 97.5 kB/s  0:00:146.1 kB/s eta 0:00:02:04m
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6/6 [pyjanitor]━ 5/6 [pyjanitor]
Note: you may need to restart the kernel to use updated packages.


In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import pyjanitor


%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

ModuleNotFoundError: No module named 'pyjanitor'

#### Import the CSV Data as a Pandas Dataframe

In [ ]:
df = pd.read_csv('data/StudentsPerformance.csv')
df = clean_names(df)

NameError: name 'clean_names' is not defined

#### Show the top 5 records

In [ ]:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


#### Shape of the dataset

In [ ]:
df.shape

(1000, 8)

### 2.2 Dataset Information
- Gender: Sex of students -> (Male/Female)
- Race/Ethnicity: Ethnicity of Students -> (Group A, B, C, D, E)
- Parental Level of Education: Parent's final education -> (High School, Some College, Associate's Degree, Bachelor's Degree,  Master's Degree)
- Lunch: Student's lunch program participation -> (standard or free/reduced)
- Test preparation Course: -> (Completed, Not Completed)
- Math Score
- Reading Score
- Writing Score 

## Data Checks to Perform
- Check missing values
- Check duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of dataset
- Check various categories present in the different categorical columns

### 3.1 Check Missing Values

In [ ]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There are no missing values in the dataset.

### 3.2 Check Duplicates

In [ ]:
if df.duplicated().sum() > 0:
    print('There are duplicate values in the dataset')
#    df.drop_duplicates(inplace=True)
else:
    print('There are no duplicate values in the dataset')

There are no duplicate values in the dataset


There are no duplicates in the dataset.

### 3.3 Check data types

In [ ]:
# Check null count and data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   gender                       1000 non-null   str  
 1   race/ethnicity               1000 non-null   str  
 2   parental level of education  1000 non-null   str  
 3   lunch                        1000 non-null   str  
 4   test preparation course      1000 non-null   str  
 5   math score                   1000 non-null   int64
 6   reading score                1000 non-null   int64
 7   writing score                1000 non-null   int64
dtypes: int64(3), str(5)
memory usage: 62.6 KB


### 3.4 Checking the number of unique values of each column

In [ ]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

### 3.5 Check statistics of data set

In [ ]:
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


#### Insight
- From the above description of numerical data, the means are very close to each other -- between 66 and 68.05
- All standard deviations are also close -- between 14.6 and 15.19
- The minimum score for math is 0; the minimum scores for writing (10) and reading (17) are much higher

### 3.6 Exploring the Data

In [58]:
# Gender
print("Categories in 'gender' column: ", end=" ")
print(df['gender'].unique())

# Race/Ethnicity
print("Categories in 'race/ethnicity' column: ", end=" ")
print(df['race/ethnicity'].unique())

# Parental Level of Education
print("Categories in 'parental level of education' column: ", end=" ")
print(df['parental level of education'].unique())

# Lunch
print("Categories in 'lunch' column: ", end=" ")
print(df['lunch'].unique())

# Test Preparation Course
print("Categories in 'test preparation course' column: ", end=" ")
print(df['test preparation course'].unique())

Categories in 'gender' column:  <StringArray>
['female', 'male']
Length: 2, dtype: str
Categories in 'race/ethnicity' column:  <StringArray>
['group B', 'group C', 'group A', 'group D', 'group E']
Length: 5, dtype: str
Categories in 'parental level of education' column:  <StringArray>
[ 'bachelor's degree',       'some college',    'master's degree',
 'associate's degree',        'high school',   'some high school']
Length: 6, dtype: str
Categories in 'lunch' column:  <StringArray>
['standard', 'free/reduced']
Length: 2, dtype: str
Categories in 'test preparation course' column:  <StringArray>
['none', 'completed']
Length: 2, dtype: str


In [59]:
# Define numerical and categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != '0']
categorical_features = [feature for feature in df.columns if df[feature].dtype == '0']

# Print Columns
print("We have {} numerical features : {}".format(len(numeric_features), numeric_features))
print("\nWe have {} categorical features : {}".format(len(categorical_features), categorical_features))

We have 8 numerical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']

We have 0 categorical features : []


### 3.X Adding Columns for "Total Score" and "Average Score"

In [60]:
# Total Score
df['total_score'] = df['math_score'] + df['reading_score'] + df['writing_score']

# Average Score
df['average_score'] = df['total_score'] / 3
df.head()

KeyError: 'math_score'

In [ ]:
reading_full = df[df['reading_score'] == 100]['average_score'].count()
print("Number of students with a perfect reading score: ", reading_full)

writing_full = df[df['writing_score'] == 100]['average_score'].count()
print("Number of students with a perfect writing score: ", writing_full)

math_full = df[df['math_score'] == 100]['average_score'].count()
print("Number of students with a perfect math score: ", math_full)

In [ ]:
# Reading score below the 25th percentile of all reading scores
reading_below_25pct = df[df['reading_score'] < df['reading_score'].quantile(0.25)]['average_score'].count()
print("Number of students with a reading score below the 25th percentile: ", reading_below_25pct)

# Writing score below the 25th percentile of all writing scores
reading_below_25pct = df[df['reading_score'] < df['reading_score'].quantile(0.25)]['average_score'].count()
print("Number of students with a reading score below the 25th percentile: ", reading_below_25pct)

# Math score below the 25th percentile of all math scores
reading_below_25pct = df[df['reading_score'] < df['reading_score'].quantile(0.25)]['average_score'].count()
print("Number of students with a reading score below the 25th percentile: ", reading_below_25pct)

#### Insights

## Data Visualization

### 4.1 Visualize average score distribution to make some conclusion
- Histogram
- Kernel Distribution Function (KDE)

#### 4.1.1 Histogram and KDE

In [ ]:
# Average score by gender
fig, axis = plt.subplots(1, 2, figsize=(15, 7))
plt.subplot(121)
sns.histplot(data=df, x='average_score', bins=30, kde=True, color='g')
plt.subplot(122)
sns.histplot(data=df, x='average_score', kde=True, hue='gender')
plt.show()

# Total score by gender

##### Insights
- Female students tend to perform better than male students

In [ ]:
# Average score by lunch program participation
plt.subplots(1, 3, figsize=(25, 6))
plt.subplot(141)
sns.histplot(data=df, x='average_score', kde=True, hue='lunch')
plt.subplot(142)
sns.histplot(data=df[df.gender=='female'], x='average_score', kde=True, hue='lunch')
plt.subplot(143)
sns.histplot(data=df[df.gender=='male'], x='average_score', kde=True, hue='lunch')
plt.show()

# Insights